Linker flavors next steps: linker features

lqd opened the pull request:
This is my understanding of the first step towards @petrochenkov's vision for the future of linker flavors, described in https://github.com/rust-lang/rust/pull/119906#issuecomment-1895693162 and the discussion that followed.

To summarize: having `Cc` and `Lld` embedded in linker flavors creates tension about naming, and a combinatorial explosion of flavors for each new linker feature we'd want to use. Linker features are an extension mechanism that is complementary to principal flavors, with benefits described in #119906.

The most immediate use of this flag would be to turn self-contained linking on and off via features instead of flavors. For example, `-Clinker-features=+/-lld` would toggle using lld instead of selecting a precise flavor, and would be "generic" and work cross-platform (whereas linker flavors are currently more tied to targets). Under this scheme, MCP510 is expected to be `-Clink-self-contained=+linker -Zlinker-features=+lld -Zunstable-options` (though for the time being, the original flags using lld-cc flavors still work).

I purposefully didn't add or document CLI support for `+/-cc`, as it would be a noop right now. I only expect that we'd initially want to stabilize `+/-lld` to begin with.

r? @petrochenkov 

You had requested that minimal churn would be done to the 230 target specs and this does none yet: the linker features are inferred from the flavor since they're currently isomorphic. We of course expect this to change sooner rather than later. 

In the future, we can allow targets to define linker features independently from their flavor, and remove the cc and lld components from the flavors to use the features instead, this actually doesn't need to block stabilization, as we discussed. 

(Best reviewed per commit)

lqd committed:
remove some unnecessary lifetimes

lqd committed:
introduce `LinkerFeatures`

They are a flexible complementary mechanism to linker flavors,
that also avoid the combinatorial explosion of mapping linking features
to actual linker flavors.

lqd committed:
add `-Z linker-features` to toggle lld on the CLI

but don't expose `+/-cc` yet

lqd committed:
test self-contained linking via `-Zlinker-features=+lld`

lqd committed:
document `-Zlinker-features` in the unstable book

petrochenkov commented:
>Also: the last commit here is purely cosmetic: 30% of the spec module was about linking so I moved these to a dedicated module. I can drop that commit if you prefer

Let's land it separately, before or after this PR.

petrochenkov commented:
What happens if `-Zlinker-features=+lld,-lld` or `-Zlinker-features=-lld -Zlinker-features=+lld`.
IIRC, for `-Clink-self-contained` we just produced an error for cases like this.

lqd commented:
> Let's land it separately, before or after this PR.

It's fine, we can land such a thing in the next PRs. I had removed it from my branch already by the time you added the other comment :)

> What happens if -Zlinker-features=+lld,-lld or -Zlinker-features=-lld -Zlinker-features=+lld.

Right now nothing spectacularly amazing, in particular I was thinking whether MCP https://github.com/rust-lang/compiler-team/issues/731 would need to be taken into account. I think so, so we could do the +/- reduction while accumulating the CLI flags (e.g. `+lld,-lld,+lld` would not result in `LinkerFeaturesCLI { enabled: lld, disabled: lld }`, but simply in "enabled: lld"). 

So I will:
- remove the cosmetic commit now
- handle these cases of duplicate flags
- remove `+/-cc` from the CLI (and documentation) only, because it's not really doing much yet and that could be confusing

lqd committed:
test duplicate features, that the last `+/-lld` on the CLI wins

lqd commented:
Updated to:
- remove the cosmetic commit, we'll land it later
- handle duplicates like `+lld,-lld` as the last feature flag wins
- only mention `-Z linker-features=+/-lld` in the code and documentation: we don't expose or mention the `cc` feature that we have until it's actually doing something on the CLI.

@rustbot ready

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -3057,24 +3093,21 @@ fn get_apple_sdk_root(sdk_name: &str) -> Result<String, errors::AppleSdkRootErro
 /// - when the self-contained linker flag is active: the build of `lld` distributed with rustc,
 /// - or any `lld` available to `cc`.
+#[instrument(level = "debug", skip(cmd, sess))]
 fn add_lld_args(
```
There's one more place where the distinction between lld and non-lld is important - `fn get_linker`, it will add `-flavor <FLAVOR>` to the command line if the linker is lld.
Right now it won't be added even if `-Clinker-features=+lld` is specified.

Ideally `fn linker_and_flavor` would return a flavor already adjusted for `-Clinker-features`, then everything would automatically work correctly after that.

lqd replied:
If that's the case then I'm not sure if we ever had `-flavor <FLAVOR>` for `-Zgcc-ld=lld` either, anything it did was way after `get_linker`.

petrochenkov replied:
`-flavor <FLAVOR>` is not needed when lld is called through cc.
It's only needed when lld is called directly, and with an "improper" name not conveying the flavor, e.g. "rust-lld".

lqd committed:
make CLI linker features influence the linker flavor

While they're isomorphic, we can flip the lld component where
applicable, so that downstream doesn't have to check both the flavor and
the linker features.

petrochenkov commented:
I'll think about the consistency checks later during stabilization.
@bors r+

bors commented:
:pushpin: Commit dee834b8a652ea867dca20192400f7331b830a7e has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov dee834b8a652ea867dca20192400f7331b830a7e -->
<!-- homu: {"type":"Approved","sha":"dee834b8a652ea867dca20192400f7331b830a7e","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit dee834b8a652ea867dca20192400f7331b830a7e with merge 7106800e16085bf1805cc8e75866ed0f0efd85d9...
<!-- homu: {"type":"BuildStarted","head_sha":"dee834b8a652ea867dca20192400f7331b830a7e","merge_sha":"7106800e16085bf1805cc8e75866ed0f0efd85d9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8673088557/job/23785459410)
Approved by: petrochenkov
Pushing 7106800e16085bf1805cc8e75866ed0f0efd85d9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8673088557/job/23785459410"},"merge_sha":"7106800e16085bf1805cc8e75866ed0f0efd85d9"} -->

rust-timer commented:
Finished benchmarking commit (7106800e16085bf1805cc8e75866ed0f0efd85d9): [comparison URL](https://perf.rust-lang.org/compare.html?start=6eaa7fb5762c9cc5e3335d0e5f1bb62ff990c623&end=7106800e16085bf1805cc8e75866ed0f0efd85d9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6eaa7fb5762c9cc5e3335d0e5f1bb62ff990c623&end=7106800e16085bf1805cc8e75866ed0f0efd85d9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6eaa7fb5762c9cc5e3335d0e5f1bb62ff990c623&end=7106800e16085bf1805cc8e75866ed0f0efd85d9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.9%  | [3.9%, 3.9%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.3% | [-4.2%, -2.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.9% | [-4.2%, 3.9%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6eaa7fb5762c9cc5e3335d0e5f1bb62ff990c623&end=7106800e16085bf1805cc8e75866ed0f0efd85d9&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=6eaa7fb5762c9cc5e3335d0e5f1bb62ff990c623&end=7106800e16085bf1805cc8e75866ed0f0efd85d9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 675.908s -> 677.551s (0.24%)
**Artifact size**: 316.10 MiB -> 316.04 MiB (-0.02%)
